### ISSUE TITLE
NCrystal in Windows

### ISSUE THREAD

**Issue by andrmor** [mctools/ncrystal#17]
I have tried to compile NCrystal in MSVC2017 but there were issues:

1. fatal error LNK1181: cannot open input file 'm.lib'
2. need to #define _USE_MATH_DEFINES  and then #include <math.h> in: NCFastConvolve.cc   and NCMath.hh
3. ncrystal_core\src\ncmath.hh(115): error C2065: 'x': undeclared identifier   (change x to a)
4. std::bind1st (several places?) -> need to add #include <functional>. By the way, the function is deprecated in c++11, and removed in c++17
5. NCMatrix.cc: swap_ranges -> std::swap_ranges   and  #include <algorithm> 
6. NCParseNCMAT.cc: #include  <cctype> for isdigit
7. defined(__GNUC__) or defined(__clang__) -> defined(__GNUC__) || defined(__clang__)

and several strange warnings:
ncparsencmat.cc(92): warning C4146: unary minus operator applied to unsigned type, result still unsigned (same in line 112, 151 and 167)

Regards,
Andrey

**Comment by tkittel** [comment]
Hi Andrey,

Thanks a lot for this super useful report, I'll make sure to include all of these fixes in the next NCrystal release. As I told you elsewhere, neither me nor @xxcai1 develop on Windows, so we rely on reports like this to make it work there as well.

**Comment by tkittel** [comment]
It is still not on github, but I just wanted to let you know that I managed to find time to (I hope) commit fixes for all of the above. Regarding item 4, we still use std::bind1st for now, but I added their removal to our todo list, since as you point out they are gone in C++17. Item 1 was the most tricky one, since I had to make some cmake magic lines which would add direct libm linkage when needed, and otherwise not. It is extra complicated since the functions in libm typically have builtin compiler support, so a small test compilation without libm might just be so unfortunate as to work. I hope I managed to cook up something that will work in all cases, time will tell.

I'll let you know here once there is something on github for you to test - I am hoping we will be able to make a new release early next week.

**Comment by tkittel** [comment]
bind1st usage has been removed as well now.

**Comment by tkittel** [comment]
Hi @andrmor,

So NCrystal v0.9.10 was just released with fixes for all of the problems you reported above :-)

However, the release includes quite a large amount of new code and code refactoring, so I can not guarantee 100% that you won't find any issues if you try to build again on Windows.

Can I ask you to try it out and report any lingering/new issues here? In case of any, we could quickly make a small follow-up release addressing those.

Cheers,
Thomas

**Comment by andrmor** [comment]
Hi Thomas!

I can compile NCrystal (I enable core + example only) on Win (no errors anymore).

However, in ANTS2 linker complains about NCrystal.dll:
> LNK1107: invalid or corrupt file: cannot read at 0x2B8
I will test at another computer and report the results.

A.

**Comment by tkittel** [comment]
Hi @andrmor,

Thanks for checking. So I will of course add the missing include. It is more worrisome that you get problems with the dll. The stackoverflow link you sent me seems to indicate that I have to add 100s of `__declspec(dllexport)` statements to our code - I certainly hope we won't have to do that :-(

But I am puzzled. I thought the only build-related change this time was the lines in the CMakeLists.txt file which concerns whether or not to link with libm. So how come you did not have any such dll issues the last time around?

Or did you change your setup in other ways?

**Comment by andrmor** [comment]
Hi Thomas,

No, last time I did not get to this step at all. Linking was failing during for NCrystal core itself.

Now this step finishes, but only dynamic link library file (dll) is generated.
And I get error during linking stage of ANTS2 when I activate support of NCrystal.

It seems you need to set output to generate a static (lib) library: I get a (not fatal, obviously) error during linking of NCrystal examples - linker complains that "cannot open input file Release/NCrystal.lib".

So I guess there are some problems (or conflicts) in cmake configuration in the case of Windows.

Andrey

**Comment by tkittel** [comment]
Hmm... I wonder if creating static rather than shared libraries would solve your issues (it would not be great for the NCrystal python module unfortunately, but one thing at a time).

It is hard for me to debug it myself without a windows dev-machine, but perhaps you could try if changing the:
```
add_library(NCrystal SHARED ${SRCS_NC})
```
To
```
add_library(NCrystal STATIC ${SRCS_NC})
```
will make a difference?

**Comment by andrmor** [comment]
Where to change it? I am not much familiar with CMake

Meantime, I have found in the studio where to change dynamic to static,
and everything works. Would be nice to make that setting in cmake.

A.

**Comment by andrmor** [comment]
Yes, confirmed - changing STATIC makes all problems go away :)

**Comment by tkittel** [comment]
Ok, so it seems we have a pretty good understanding of where the problem comes from then.

I have just put a candiate for release 0.9.11 in the develop branch here at github, in which I bit the bullet and tried to add a new header `ncapi.h` which provides a new macro NCRYSTAL_API which I have then attempted to put in all the exported header files on all functions and classes. Hopefully the preprocessor stuff I created in `ncapi.h` will do the right thing on Windows and you will get your dll's without any STATIC hacks (which has the nasty side-effect of making the python interface not work).  As you can see if you look inside ncapi.h, on windows it requires that the flag `-DNCrystal_EXPORTS` is set when building the ncrystal library, but not when building the client applications. Supposedly CMake will take care of that, but if you are not using CMake, you have to make sure to build with that flag.

Having said that, I obviously have not been able to try it out on a windows machine. Instead, I ran gcc on linux with `-fvisibility=hidden` to at least make sure I didn't forget the NCRYSTAL_API macros anywhere important (it did give me missing symbol failures at first until I added it a few more places...).

Oh, and I didn't add any NCRYSTAL_API stuff for the NCrystal-Geant4 interface library yet, but as far as I know that doesn't affect you anyway.

If you have time, could you try out the code currently at the head of the develop branch and see if your windows build goes better now?

**Comment by tkittel** [comment]
Great! NCrystal v0.9.11 is now out with these changes. Thanks a lot for the help with this. I'll close this issue then, please let us know next time something like this comes up :-)

**Comment by tkittel** [comment]
@andrmor note that we expect NCrystal to properly support windows later this year, with PyPI/conda packages available (tracked at #154).
